### PR TITLE
Add support for the modulo operator

### PR DESCRIPTION
**Pull request by tirkarp**
https://gitlab.com/purdue-informatics/variate/variate-solution-engine/-/issues/58

### CODE REVIEW

**Comment by gavin-brown** [comment]
@tirkarp I need to actually run it but the code looks great from what I can see!

File: tests/mod_test.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+from .context import assert_equal
+import pytest
+from sympy import Symbol, Rational, Mod, sqrt, nsimplify, pi, GoldenRatio
+from sympy.physics.units import hbar
+
+x = Symbol('x', real=True)
+y = Symbol('y', real=True)
+
+
+def test_mod_usual():
```
Comment (lily514) [inline tests/mod_test.py:10]: @tirkarp this is an awesome test suite! nice job!

Comment (lily514) [reply tests/mod_test.py:10]: We might consider adding a test to the all_bad_test.py to check that we get the matrix operand error.

File: tests/mod_test.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def test_mod_float():
+    assert_equal("0.41\\mod 2", Mod(Rational('0.41'), 2))
+    assert_equal("143E-13\\mod 21", Mod(Rational('143E-13'), 21))
+    assert_equal("-9.80665\\mod 9.80665", Mod(-9.80665, 9.80665))
+    assert_equal("0.0000923423\\mod -8341.234802909", nsimplify(Mod(0.0000923423, -8341.234802909)))
+    assert_equal("\\sqrt{5}\\mod \\sqrt{2}", Mod(sqrt(5), sqrt(2)))
+    assert_equal("987\\mod \\pi", Mod(987, pi))
+    assert_equal("\\pi\\mod ((1+\\sqrt{5})/2) ", Mod(pi, GoldenRatio))
```
Comment (lily514) [inline tests/mod_test.py:52]: @tirkarp now that I'm running pytest, I'm getting a failure on this one. 
`actual = -sqrt(5)/2 - 1/2 + pi, expected = pi - GoldenRatio`

File: tests/mod_test.py (hunk 1 above)
Comment (tirkarp) [reply tests/mod_test.py:10]: Added.

File: tests/mod_test.py (hunk 2 above)
Comment (tirkarp) [reply tests/mod_test.py]: That should've been caught on my end, but it passed on my machine. Hopefully it's fixed now with the new commit.

File: latex2sympy.py [hunk 3]
```diff
@@ -250,6 +250,13 @@ def convert_mp(mp):
             return sympy.MatMul(lh, sympy.Pow(rh, -1, evaluate=False), evaluate=False)
         else:
             return sympy.Mul(lh, sympy.Pow(rh, -1, evaluate=False), evaluate=False)
+    elif mp.CMD_MOD():
+        lh = convert_mp(mp_left)
+        rh = convert_mp(mp_right)
+        if rh.is_Matrix:
+            raise Exception("Cannot perform modulo operation with a matrix as an operand")
+        else:
+            return sympy.Mod(lh, rh)
```
Comment (builtbywill) [inline latex2sympy.py:259]: @tirkarp I believe we will want to pass `evaluate=False` into Mod, in order to preserve the expression

File: tests/mod_test.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def test_mod_float():
+    assert_equal("0.41\\mod 2", Mod(Rational('0.41'), 2))
+    assert_equal("143E-13\\mod 21", Mod(Rational('143E-13'), 21))
+    assert_equal("-9.80665\\mod 9.80665", Mod(-9.80665, 9.80665))
+    assert_equal("0.0000923423\\mod -8341.234802909", nsimplify(Mod(0.0000923423, -8341.234802909)))
+    assert_equal("\\sqrt{5}\\mod \\sqrt{2}", Mod(sqrt(5), sqrt(2)))
+    assert_equal("987\\mod \\pi", Mod(987, pi))
+    assert_equal("\\pi\\mod ((1+\\sqrt{5})/2) ", Mod(pi, GoldenRatio), symbolically=True)
```
Comment (builtbywill) [inline tests/mod_test.py:52]: @tirkarp This test is failing for me

Comment (tirkarp) [reply tests/mod_test.py]: 🤔 Weird. I still can't get that failure reproduced. Can you check if it is fixed now?

Comment (builtbywill) [reply tests/mod_test.py]: @tirkarp yeah it does pass now

File: tests/mod_test.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    assert_equal("0.41\\mod 2", Mod(Rational('0.41'), 2))
+    assert_equal("143E-13\\mod 21", Mod(Rational('143E-13'), 21))
+    assert_equal("-9.80665\\mod 9.80665", Mod(-9.80665, 9.80665))
+    assert_equal("0.0000923423\\mod -8341.234802909", nsimplify(Mod(0.0000923423, -8341.234802909)))
+    assert_equal("\\sqrt{5}\\mod \\sqrt{2}", Mod(sqrt(5), sqrt(2)))
+    assert_equal("987\\mod \\pi", Mod(987, pi))
+    assert_equal("\\pi\\mod ((1+\\sqrt{5})/2)", Mod(pi, nsimplify(GoldenRatio)), symbolically=True)
+
+    # a number modded with any rational number where the numerator is 1 is always zero
+    # TODO: incidentally, `Mod(1234, Rational('1E-29'))` gives a wrong value
```
Comment (builtbywill) [inline tests/mod_test.py:55]: @tirkarp If instead you compare it with `Mod(1234, Rational('1E-29'), evaluate=False)`, it passes

without the `evaluate=False` the expected value is going to automatically evaluated to a number at a given precision, which won't then match the unevaluated expression that is parsed

Comment (tirkarp) [reply tests/mod_test.py]: ah thank you. fixed.

**Pull request merged**